fendermint/vm/interpreter/src/fvm/gas/actor.rs

```diff
@@ -69,13 +82,56 @@ impl ActorGasMarket {
         Ok(Self {
             block_gas_limit: reading.block_gas_limit,
             block_gas_used: 0,
+            constant_update: None,
         })
     }
 
+    pub fn process_consensus_update(&self, update: &mut ConsensusBlockUpdate) {
+        if let Some(ref set_constant) = self.constant_update {
+            update.process_block_size(set_constant.block_gas_limit);
+        }
+    }
+
     pub fn commit<E: Executor>(
         &self,
         executor: &mut E,
         block_height: ChainEpoch,
+    ) -> anyhow::Result<()> {
+        self.commit_constants(executor, block_height)?;
+        self.commit_utilization(executor, block_height)
+    }
+
+    fn commit_constants<E: Executor>(
+        &self,
+        executor: &mut E,
+        block_height: ChainEpoch,
+    ) -> anyhow::Result<()> {
+        let Some(ref constants) = self.constant_update else {
+            return Ok(());
+        };
+
+        let msg = FvmMessage {
+            from: system::SYSTEM_ACTOR_ADDR,
+            to: GAS_MARKET_ACTOR_ADDR,
+            sequence: block_height as u64,
+            // exclude this from gas restriction
+            gas_limit: i64::MAX as u64,
+            method_num: fendermint_actor_gas_market::Method::SetConstants as u64,
+            params: fvm_ipld_encoding::RawBytes::serialize(constants)?,
+            value: Default::default(),
+            version: Default::default(),
+            gas_fee_cap: Default::default(),
+            gas_premium: Default::default(),
+        };
+        self.call_fvm(msg, executor)?;
+
+        Ok(())
+    }
```

Comment:
We don't need to handle constant changes at all, these happen entirely as standard chain messages, right? Or that's what I thought?

The only thing we need to be careful with is not changing constants _in the middle of block execution_. So I think we'll need a message selection / prepare proposal / process proposal rule to enforce messages directed to the gas market actor to always be at the top of the block.

Does that make sense?

So given a data structure:

```rust
struct MessageOrderRule {
    from: Option<Address>,
    to: Option<Address>,
    precedence: i32,
}
```

We could have a lazy static:

```
lazy_static! {
    pub static ref MESSAGE_ORDER_POLICY: Vec<MessageOrderRule> = vec![
        MessageOrderRule{from: None, to: Some(Address::from_id(GAS_MARKET_ID), precedence: 1000},
}
```

And then we'd reoder messages during selection to satisfy the policy?

Actually, we could have this policy be a trait with methods to sort (prepare proposal) and validate order (process proposal).

Reply:
I think a trait is better, more general and flexible. Just holding a vec of `MessageOrderRule` traits implementations.